File: 3413.maximum-coins-from-k-consecutive-bags.cpp
#
# @lc app=leetcode id=3413 lang=cpp
#
# [3413] Maximum Coins From K Consecutive Bags
#
# @lc code=start
class Solution {
public:
    long long maximumCoins(vector<vector<int>>& coins, int k) {
        sort(coins.begin(), coins.end());
        
        long long maxCoins = 0;
        int n = coins.size();
        
        for (int i = 0; i < n; i++) {
            // Try window starting at coins[i][0]
            long long start = coins[i][0];
            long long end = start + k - 1;
            long long sum = 0;
            
            for (int j = i; j < n && coins[j][0] <= end; j++) {
                long long l = coins[j][0];
                long long r = coins[j][1];
                long long c = coins[j][2];
                
                long long overlapLeft = max(l, start);
                long long overlapRight = min(r, end);
                
                if (overlapLeft <= overlapRight) {
                    sum += (overlapRight - overlapLeft + 1) * c;
                }
            }
            
            maxCoins = max(maxCoins, sum);
            
            // Try window ending at coins[i][1]
            end = coins[i][1];
            start = end - k + 1;
            sum = 0;
            
            for (int j = 0; j < n && coins[j][0] <= end; j++) {
                long long l = coins[j][0];
                long long r = coins[j][1];
                long long c = coins[j][2];
                
                long long overlapLeft = max(l, start);
                long long overlapRight = min(r, end);
                
                if (overlapLeft <= overlapRight) {
                    sum += (overlapRight - overlapLeft + 1) * c;
                }
            }
            
            maxCoins = max(maxCoins, sum);
        }
        
        return maxCoins;
    }
};
# @lc code=end

File: 3413.maximum-coins-from-k-consecutive-bags.py
#
# @lc app=leetcode id=3413 lang=python3
#
# [3413] Maximum Coins From K Consecutive Bags
#
# @lc code=start
from typing import List
from bisect import bisect_left, bisect_right
class Solution:
    def maximumCoins(self, coins: List[List[int]], k: int) -> int:
        # Collect all unique segment endpoints
        events = []
        for l, r, c in coins:
            events.append((l, c))
            events.append((r + 1, -c))
        events.sort()
        # Build position and prefix sum arrays
        pos = []
        coin_at = []
        cur = 0
        prev = None
        for p, delta in events:
            if prev is not None and p != prev:
                pos.append(prev)
                coin_at.append(cur)
            cur += delta
            prev = p
        # Add last position if needed
        if prev is not None:
            pos.append(prev)
            coin_at.append(cur)
        n = len(pos)
        # Build prefix sum of coins over positions
        prefix = [0]
        for i in range(n - 1):
            segment_len = pos[i+1] - pos[i]
            prefix.append(prefix[-1] + coin_at[i] * segment_len)
        max_coins = 0
        # Slide window of size k
        for i in range(n - 1):
            window_start = pos[i]
            window_end = window_start + k
            # Find the rightmost index where pos[j] < window_end
            j = bisect_left(pos, window_end, i, n) - 1
            total = prefix[j+1] - prefix[i]
            # Add partial segment if needed
            overlap = max(0, window_end - pos[j])
            if j < n - 1 and overlap > 0:
                total += coin_at[j] * overlap
            max_coins = max(max_coins, total)
        return max_coins
# @lc code=end